bunkering/BunkeringMDP.py

```diff
@@ -0,0 +1,260 @@
+            return 1086
+
+
+class BunkeringAction(Enum):
+    pass
+
+    def speed(self):
+        return self.value[0]
+
+    def refuel_amount(self):
+        return self.value[1]
+
+
+def setBunkeringActionAttributes():
+    for i, action in enumerate(POSSIBLE_ACTIONS):
+        name = f'A_{action[0]}_{action[1]}'
+        extend_enum(BunkeringAction, name, action)
+
+
+class BunkeringPosition():
+    def __init__(self, port: Port, fuel_amount) -> None:
+        self.port = port
+        self.fuel_amount = fuel_amount
+
+    def __eq__(self, __o: object) -> bool:
+        if isinstance(__o, BunkeringPosition):
+            return __o.port == self.port and __o.fuel_amount == self.fuel_amount
+        return False
+
+    def __str__(self) -> str:
+        return f'{self.port}, {self.fuel_amount}'
+
+
+class BunkeringReward(BunkeringPosition):
+    def __init__(self, port: Port, fuel_amount, arrival_time, min_fuel_allowance, kappa=KAPPA, kappa2=KAPPA2):
+        super().__init__(port, fuel_amount)
+        self.arrival_time = arrival_time
+        self.kappa = kappa
+        self.kappa2 = kappa2
+        self.min_fuel_allowance = min_fuel_allowance
+        self.reward = 0
+
+    def calculate_reward(self):
+        if self.port.arrival_allowance_time() >= self.port.arrival_time:
+            pass
+        else:
+            self.reward -= (self.port.arrival_time - self.port.arrival_allowance_time()) * self.kappa
+
+        if self.fuel_amount < self.min_fuel_allowance:
+            self.reward -= (self.min_fuel_allowance - self.fuel_amount) * self.kappa2
+        if self.fuel_amount <= 0:
+            self.reward -= math.inf
+
+
+class BunkeringState(BunkeringPosition):
+    def __init__(self, port: Port, fuel_amount, speed):
+        super().__init__(port, fuel_amount)
+
+        self.is_terminal = port.is_terminal()
+        self.speed = speed
+
+    def is_action_valid(self, action: BunkeringAction, current_fuel_size) -> bool:
+        if action.speed() <= MAX_SPEED and action.refuel_amount() + current_fuel_size <= MAX_FUEL_SIZE and \
+                not self.is_terminal:
+            return True
+        else:
+            return False
+
+    def resolve_action(self, action: BunkeringAction, current_fuel_size) -> Optional[
+        BunkeringState]:
+        if self.is_action_valid(action, current_fuel_size):
+            return BunkeringState(port=self.port.next_port(), fuel_amount=current_fuel_size + action.refuel_amount(),
+                                  speed=action.speed())
+        return None
+
+
+class BunkeringMDP(MDP[BunkeringState, BunkeringAction]):
+    def __init__(self, TEU, minimum_allowable_fuel_percentage: float = 0.9):
+        self.minimum_allowable_fuel_percentage = minimum_allowable_fuel_percentage
+        self.starting_state = BunkeringState(Port(0, arrival_time=0), fuel_amount=MAX_FUEL_SIZE, speed=INITIAL_SPEED)
+        self.current_fuel = MAX_FUEL_SIZE
+        self.TEU = TEU
```

Comment:
By convention upper case letters such as `TEU` should be global, and not contained in side an object.